page_test.go

```diff
@@ -52,93 +51,24 @@ func TestPageMarshalling(t *testing.T) {
 	}
 }
 
-// BenchmarkPageMarshalling benchmarks the marshalling of pages in a worst case scenario where all
-// pages are processed in the same thread
-func BenchmarkPageMarshalling(b *testing.B) {
-	nextPage := page{
-		offset: 12345,
-	}
-	currentPage := page{
-		nextPage:            &nextPage,
-		offset:              4096,
-		transactionNumber:   42,
-		payload:             fastrand.Bytes(4096 - 64), // ensure marshalled size is 4096 bytes
-		pageStatus:          pageStatusComitted,
-		transactionChecksum: [crypto.HashSize]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+// BenchmarkPageWriteTo benchmarks the writeTo method of page.
+func BenchmarkPageWriteTo(b *testing.B) {
+	p := page{
+		offset:            4096,
+		transactionNumber: 42,
+		payload:           fastrand.Bytes(maxPayloadSize), // ensure marshalled size is 4096 bytes
+		pageStatus:        pageStatusComitted,
+		nextPage: &page{
+			offset: 12345,
+		},
 	}
-
-	// Create around 400mb of pages
-	pages := 100000
-
-	// Start timer after the test was set up
+	fastrand.Read(p.transactionChecksum[:])
 	b.ResetTimer()
+	b.ReportAllocs()
 
-	// Marshal and unmarshal pagesPerThread pages
-	for i := 0; i < pages; i++ {
-		marshalledBytes, err := currentPage.Marshal()
-		if err != nil {
+	for i := 0; i < b.N; i++ {
+		if err := p.writeTo(ioutil.Discard); err != nil {
 			b.Fatalf("Failed to marshal the page %v", err)
 		}
-
-		pageRestored := page{}
-		_, err = unmarshalPage(&pageRestored, marshalledBytes)
-		if err != nil {
-			b.Fatalf("Failed to unmarshal the page %v", err)
-		}
-	}
-}
-
-// BenchmarkPageMarshallingParallel benchmarks the marshalling of pages in a best case scenario where all
-// pages are evenly distributed across cpu cores
-func BenchmarkPageMarshallingParallel(b *testing.B) {
```

Comment:
It's probably fine to have a parallel benchmark, even there's no strict concurrency here. For example, sometimes your dependencies have hidden concurrency (like `crypto/rand`) where they will bottleneck an expensive part of the process to a single thread, and it's good to know you have that bottleneck.

Reply:
I think parallel benchmarks are most useful when they test the highest-level function that gets parallelized. For example, in the WAL we want to know what happens to its performance when writing multiple transactions in parallel. Parallel benchmarks of the individual functions aren't as useful, because they don't reflect the real-world conditions in which the function will be called (i.e. there's no way to tell whether a particular function will bottleneck the others, because you don't know how frequently it's called in relation to them). So I'd prefer to the parallel benchmarks to be at a high level in order to more accurately diagnose where the bottlenecks are.